File: EMTImageCacheManager/EMTImageCacheManager.m
//
//  EMTImageCacheManager.m
//
//  Created by Hironobu Kimura on 2015/05/01.
//  Copyright (c) 2015 emotionale. All rights reserved.
//

#import "EMTImageCacheManager.h"

@implementation EMTImageCacheManager {
    NSMutableArray *keyInformations;
}

static EMTImageCacheManager *_instance = nil;

+ (instancetype)instance {
    static dispatch_once_t onceToken;
    dispatch_once(&onceToken, ^{
        _instance = [[EMTImageCacheManager alloc] init];
    });
    return _instance;
}

- (void)prepareOrderedCacheInformations {
    if (!keyInformations) {
       keyInformations = [self getOrderedKeyInformations];
    }
}

- (NSMutableArray *)getOrderedKeyInformations {
    
    NSError *error = nil;
    NSRegularExpression *regularExpression = [NSRegularExpression regularExpressionWithPattern:@"^cache([0-9.]+)_(.+)$" options:0 error:&error];
    
    NSMutableArray *keyInfos = [NSMutableArray array];
    NSDictionary *caches = [WKInterfaceDevice currentDevice].cachedImages;
    for (NSString *key in caches) {
        NSArray *matches = [regularExpression matchesInString:key options:0 range:NSMakeRange(0, [key length])];
        if ([matches count] != 0) {
            NSTextCheckingResult *result = (NSTextCheckingResult *)matches[0];
            NSString *dateString = [key substringWithRange:[result rangeAtIndex:1]];
            NSString *name = [key substringWithRange:[result rangeAtIndex:2]];
            NSDictionary *dict = @{ @"date":[NSDate dateWithTimeIntervalSince1970:[dateString doubleValue]], @"key":key, @"name":name };
            [keyInfos addObject:dict];
        }
    }
    
    NSArray *sortedKeyInfos = [keyInfos sortedArrayUsingComparator:^(id obj1, id obj2) {
                                return [obj1[@"date"] compare:obj2[@"date"]];
                                }];
    
    return [sortedKeyInfos mutableCopy];
}

- (NSString *)addOrderedCachedImageWithData:(NSData *)imageData name:(NSString *)name {

    if (!imageData || !name) return nil;
    
    [self prepareOrderedCacheInformations];
    
    [self removeOrderedCachedImageForName:name];
    
    NSString *key = [self saveOrderedCachedImageWithData:imageData name:name];
    if (key) {
        return key;
    }

    NSInteger dataSize = [imageData length];

    NSMutableIndexSet *indexes = [NSMutableIndexSet indexSet];
    
    for (NSUInteger i = 0; i < [keyInformations count]; i++) {
        NSString *imageKeyToRemove = keyInformations[i][@"key"];
        NSNumber *num = [WKInterfaceDevice currentDevice].cachedImages[imageKeyToRemove];
        dataSize -= [num integerValue];
        [[WKInterfaceDevice currentDevice] removeCachedImageWithName:imageKeyToRemove];
        [indexes addIndex:i];
        if (dataSize <= 0) break;
    }
    if ([indexes count] != 0) {
        [keyInformations removeObjectsAtIndexes:indexes];
    }
    
    return [self saveOrderedCachedImageWithData:imageData name:name];
}

- (NSString *)getOrderedCacheKeyForName:(NSString *)name {
    
    [self prepareOrderedCacheInformations];
    
    NSInteger index = [self indexOfOrderdKeyWithName:name];
    if (index == -1) {
       return nil;
    }
    return keyInformations[index][@"key"];
}

- (NSString *)saveOrderedCachedImageWithData:(NSData *)imageData name:(NSString *)name {
    
    NSDate *date = [NSDate date];
    NSString *key = [NSString stringWithFormat:@"cache%f_%@", [date timeIntervalSince1970], name];
    
    if ([[WKInterfaceDevice currentDevice] addCachedImageWithData:imageData name:key]) {
        [keyInformations addObject:@{ @"date":date, @"key":key, @"name":name }];
        return key;
    }
    return nil;
}

- (NSInteger)indexOfOrderdKeyWithName:(NSString *)name {
    
    __block NSInteger index = -1;
    [keyInformations enumerateObjectsUsingBlock:^(NSDictionary *dict, NSUInteger idx, BOOL *stop) {
        if ([dict[@"name"] isEqualToString:name]) {
            index = idx;
            *stop = YES;
        }
    }];
    return index;
}

- (void)removeOrderedCachedImageForName:(NSString *)name {
    
    [self prepareOrderedCacheInformations];
    
    NSInteger removeIndex = [self indexOfOrderdKeyWithName:name];
    if (removeIndex != -1) {
        [[WKInterfaceDevice currentDevice] removeCachedImageWithName:keyInformations[removeIndex][@"key"]];
        [keyInformations removeObjectAtIndex:removeIndex];
    }
}

- (void)removeAllOrderedCachedImage {
    
    [self prepareOrderedCacheInformations];
    
    [keyInformations enumerateObjectsUsingBlock:^(NSDictionary *dict, NSUInteger idx, BOOL *stop) {
        [[WKInterfaceDevice currentDevice] removeCachedImageWithName:dict[@"key"]];
    }];
    [keyInformations removeAllObjects];
}

@end

File: EMTImageCacheManager/EMTImageCacheManager.swift
//
//  EMTImageCacheManager.swift
//
//  Created by Hironobu Kimura on 2015/05/01.
//  Copyright (c) 2015 emotionale. All rights reserved.
//

import WatchKit

public class EMTImageCacheManager: NSObject {

    typealias keyInformationDict = Dictionary<String, AnyObject>
    
    var keyInformations: [keyInformationDict]?
    
    public class var instance: EMTImageCacheManager {
        struct Singleton {
            static let instance = EMTImageCacheManager()
        }
        return Singleton.instance
    }
    
    override init() {
        super.init()
    }
    
    public func prepareOrderedCacheInformations() {
        if (keyInformations != nil) {
            return
        }
        keyInformations = self.getOrderedKeyInformations()
    }
    
    private func getOrderedKeyInformations() -> [keyInformationDict] {
    
        let regularExpression = NSRegularExpression(pattern: "^cache([0-9.]+)_(.+)$", options: nil, error: nil)!
        
        var keyInfos = [keyInformationDict]()
        
        for (key, value) in WKInterfaceDevice.currentDevice().cachedImages {
            let keyString = key as! NSString
            let matches = regularExpression.matchesInString(keyString as String, options: nil, range: NSMakeRange(0, keyString.length))
            if (matches.count != 0) {
                let result = matches[0] as! NSTextCheckingResult
                let dateString = keyString.substringWithRange(result.rangeAtIndex(1)) as NSString
                let name = keyString.substringWithRange(result.rangeAtIndex(2))
                let dict = [ "date":NSDate(timeIntervalSince1970:dateString.doubleValue), "key":key, "name":name ]
                keyInfos.append(dict)
            }
        }

        let sortedKeyInfos = sorted(keyInfos) {
            (dict1 : keyInformationDict, dict2: keyInformationDict) in
            
            let date1 = dict1["date"] as! NSDate
            let date2 = dict2["date"] as! NSDate
            let result = date1.compare(date2)
            return (result == NSComparisonResult.OrderedDescending)
        }

        return sortedKeyInfos
    }
    
    public func addOrderedCachedImageWithData(data: NSData, name: String) -> String? {
    
        if (data.length == 0 || name.lengthOfBytesUsingEncoding(NSUTF8StringEncoding) == 0) {
            return nil
        }
    
        self.prepareOrderedCacheInformations()
        self.removeOrderedCachedImageForName(name)
        
        if let key = self.saveOrderedCachedImageWithData(data, name:name) {
            return key
        }
        
        var dataSize = data.length
        
        var indexes = NSMutableIndexSet()
        
        for (index, element) in enumerate(keyInformations!) {
            let dict = element as keyInformationDict

            let imageKeyToRemove = dict["key"] as! String
            if let num = WKInterfaceDevice.currentDevice().cachedImages[imageKeyToRemove] as? NSNumber {
                dataSize -= num.integerValue
                WKInterfaceDevice.currentDevice().removeCachedImageWithName(imageKeyToRemove)
                indexes.addIndex(index)
                if (dataSize <= 0) {
                    break
                }
            }
            else {
                continue
            }
            
        }
        if (indexes.count != 0) {
            keyInformations!.removeAtIndexes(indexes)
        }
        
        return self.saveOrderedCachedImageWithData(data, name:name)
    }
    
    private func saveOrderedCachedImageWithData(data: NSData, name: String) -> String? {
        let date = NSDate()
        let key = NSString(format: "cache%f_%@", date.timeIntervalSince1970, name as NSString) as String
        if (WKInterfaceDevice.currentDevice().addCachedImageWithData(data, name:key)) {
            let dict = [ "date":date, "key":key, "name":name ] as keyInformationDict
            keyInformations!.append(dict)
            return key
        }
        return nil
    }

    public func getOrderedCacheKeyForName(name: String) -> String? {
        self.prepareOrderedCacheInformations()
        let index = self.indexOfOrderdKeyForName(name)
        if (index != -1) {
            return keyInformations![index]["key"] as? String
        }
        return nil
    }
    
    public func removeOrderedCachedImageForName(name: String) {
        self.prepareOrderedCacheInformations()
        let removeIndex = self.indexOfOrderdKeyForName(name)
        if (removeIndex != -1) {
            let key = keyInformations![removeIndex]["key"] as! String
            WKInterfaceDevice.currentDevice().removeCachedImageWithName(key)
            keyInformations!.removeAtIndex(removeIndex)
        }
    }
    
    public func removeAllOrderedCachedImage() {
        self.prepareOrderedCacheInformations()
        for (index, element) in enumerate(keyInformations!) {
            let dict = element as keyInformationDict
            let imageKeyToRemove = dict["key"] as! String
            WKInterfaceDevice.currentDevice().removeCachedImageWithName(imageKeyToRemove)
        }
        keyInformations!.removeAll(keepCapacity: false)
    }
    
    private func indexOfOrderdKeyForName(name: String) -> Int {
        for (index, element) in enumerate(keyInformations!) {
            let dict = element as keyInformationDict
            if (dict["name"] as! String == name) {
                return index
            }
        }
        return -1
    }
    
}

extension Array {
    mutating func removeAtIndexes(indexes: NSIndexSet) {
        for var i = indexes.lastIndex; i != NSNotFound; i = indexes.indexLessThanIndex(i) {
            self.removeAtIndex(i)
        }
    }
}


